src/os/bluestore/BlueStore.cc

```diff
@@ -532,6 +532,184 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
+        // collocated uncompressed blob already exists
+    	--blob_info_counted->expected_allocations;
+        dout(30) << __func__  << " --expected_allocations:"
+		 << alloc_unit_start << dendl;
+      }
+      used_alloc_unit = alloc_unit_end;
+      blob_info_counted = nullptr;
+    }
+  }
+
+  for (auto b_it = affected_blobs.begin();
+       b_it != affected_blobs.end();
+       ++b_it) {
+    BlobInfo& bi = b_it->second;
+    if (bi.referenced_bytes == 0) {
+      auto& cur_collect_candidates = bi.collect_candidates;
+      uint64_t len_on_disk = b_it->first->get_blob().get_ondisk_length();
+      bi.expected_for_release +=
+        ROUND_UP_TO(len_on_disk, min_alloc_size) / min_alloc_size;
+
+      dout(30) << __func__ << " " << *(b_it->first)
+               << " expected4release=" << bi.expected_for_release
+               << " expected_allocations=" << bi.expected_allocations
+               << dendl;
+      if (bi.expected_for_release >= bi.expected_allocations) {
+        if (extents_to_collect.empty()) {
+          extents_to_collect = std::move(cur_collect_candidates);
+        } else {
+          extents_to_collect.reserve(
+            extents_to_collect.size() + cur_collect_candidates.size());
+          std::move(std::begin(cur_collect_candidates),
+                    std::end(cur_collect_candidates),
+                    std::back_inserter(extents_to_collect));
+        }
+        expected_for_release += bi.expected_for_release;
+        expected_allocations += bi.expected_allocations;
+      }
+    }
+  }
+}
+
+int64_t BlueStore::GarbageCollector::estimate(
+  uint64_t start_offset,
+  uint64_t length,
+  const BlueStore::ExtentMap& extent_map,
+  const BlueStore::extent_map_t& old_extents,
+  uint64_t min_alloc_size)
+{
+
+  affected_blobs.clear();
+  extents_to_collect.clear();
+  used_alloc_unit = boost::optional<uint64_t >();
+  blob_info_counted = nullptr;
+
+  gc_start_offset = start_offset;
+  gc_end_offset = start_offset + length;
+
+  uint64_t end_offset = start_offset + length;
+
+  for (auto it = old_extents.begin(); it != old_extents.end(); ++it) {
+    Blob* b = it->blob.get();
+    if (b->get_blob().is_compressed()) {
+
+      // update gc_start_offset/gc_end_offset if needed
+      gc_start_offset = min(gc_start_offset, (uint64_t)it->blob_start());
+      gc_end_offset = max(gc_end_offset, (uint64_t)it->blob_end());
+
+      auto o = it->logical_offset;
+      auto l = it->length;
+
+      uint64_t ref_bytes = b->get_referenced_bytes();
+      assert(l <= ref_bytes);
+      // micro optimization to bypass blobs that have no more references
+      if (ref_bytes != l) {
+        dout(30) << __func__ << " affected_blob:" << *b
+                 << " unref 0x" << std::hex << o << "~" << l
+                 << std::dec << dendl;
+	BlobInfo& bi = affected_blobs.emplace(b, BlobInfo(ref_bytes)).first->second;
+        bi.referenced_bytes -= l;
```

Comment:
emplace() call might return previously existed entry from the map if any instead of inserting a new one. Hence we need to decrement counter afterwards,